amy/workshops/templatetags/training_progress.py

```diff
@@ -53,3 +57,42 @@ def progress_description(progress):
     )
     text = text[0].upper() + text[1:]
     return mark_safe(text)
+
+
+@register.simple_tag
+def checkout_deadline(start_date):
+    # we allow 90 days for checkout
+    return start_date + timedelta(days=90)
+
+
+@register.simple_tag
+def progress_trainee_view(progress: TrainingProgress) -> str:
+    assert isinstance(progress, TrainingProgress)
+
+    # state: follow our internal choices
+    # except for Welcome Session
+    # as 'passed' implies assessment, but you just have to show up
+    state_display = progress.get_state_display().lower()
+    if state_display == "passed" and progress.requirement.name == "Welcome Session":
+        state_display = "completed"
+
+    # date: show event dates for training, most recent update date otherwise
+    date = progress.event.end if progress.event else progress.last_updated_at
+
+    # notes: show notes if state is failed or asked to repeat
+    # TODO: implement a separate field for these notes
+    # notes = (
+    #     f"<p>Administrator comments: {progress.notes}</p>"
+    #     if progress.state in ["f", "a"]
+    #     else ""
+    # )
+
+    # put it all together
+    text = (
+        f'<p class="text-{progress_state_class(progress.state)}"> '
+        f"{progress.requirement.name} {state_display} "
+        f'as of {date.strftime("%B %d, %Y")}.</p>'
+        # f'{notes}' # TODO
+    )
+
+    return mark_safe(text)
```

Comment:
Feels like instead of doing a template tag with some conditions you could do a template instead.

Reply:
Good point, done.